File: 0_1.baekjoon/graph/java/src/BOJ16932.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Set;
import java.util.StringTokenizer;

public class BOJ16932 {
    private static int N, M;
    private static int[][] board;
    private static int[][] group;
    private static int[] size;
    private static int[][] d = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        board = new int[N][M];
        group = new int[N][M];
        size = new int[N * M + 1];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < M; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        System.out.println(solve());
    }

    private static int solve() {
        int answer = 0;
        grouping();
        Set<Integer> set = new HashSet<>();
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (board[i][j] != 0) {
                    continue;
                }

                int result = 1;

                if (i - 1 >= 0) {
                    result += size[group[i - 1][j]];
                    set.add(group[i - 1][j]);
                }

                if (i + 1 < N && !set.contains(group[i + 1][j])) {
                    result += size[group[i + 1][j]];
                    set.add(group[i + 1][j]);
                }

                if (j - 1 >= 0 && !set.contains(group[i][j - 1])) {
                    result += size[group[i][j - 1]];
                    set.add(group[i][j - 1]);
                }

                if (j + 1 < M && !set.contains(group[i][j + 1])) {
                    result += size[group[i][j + 1]];
                }

                answer = Math.max(answer, result);
                set.clear();
            }
        }

        return answer;
    }

    private static void grouping() {
        boolean[][] visited = new boolean[N][M];
        int seq = 1;

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (visited[i][j] || board[i][j] == 0) {
                    continue;
                }
                groupingHelp(i, j, seq++, visited);
            }
        }
    }

    private static void groupingHelp(int row, int col, int seq, boolean[][] visited) {
        Queue<int[]> q = new LinkedList<>();
        visited[row][col] = true;
        q.add(new int[]{row, col});

        while (!q.isEmpty()) {
            row = q.peek()[0];
            col = q.poll()[1];

            group[row][col] = seq;
            size[seq]++;

            for (int i = 0; i < 4; i++) {
                int nr = row + d[i][0], nc = col + d[i][1];

                if (nr < 0 || nr >= N || nc < 0 || nc >= M) {
                    continue;
                }
                if (board[nr][nc] == 0) {
                    continue;
                }
                if (visited[nr][nc]) {
                    continue;
                }
                visited[nr][nc] = true;
                q.add(new int[]{nr, nc});
            }
        }
    }
}

File: 0_1.baekjoon/graph/kotlin/src/BOJ16932.kt
import java.util.*
import kotlin.collections.HashSet
import kotlin.math.max

private var N = 0
private var M = 0
private lateinit var board: Array<IntArray>
private lateinit var group: Array<IntArray>
private lateinit var size: IntArray
private val d = arrayOf(arrayOf(1, 0), arrayOf(0, 1), arrayOf(-1, 0), arrayOf(0, -1))
fun main() = with(System.`in`.bufferedReader()) {
    val (n, m) = readLine().split(" ").map { it.toInt() }
    N = n
    M = m

    board = Array(n) {
        readLine().split(" ").map { it.toInt() }.toIntArray()
    }

    group = Array(n) { IntArray(m) }
    size = IntArray(n * m + 1)
    print(solve())
}

private fun solve(): Int {
    grouping()
    var answer = 0
    val set = HashSet<Int>()

    for (i in 0 until N) {
        for (j in 0 until M) {
            if (board[i][j] == 1) continue
            var result = 1

            if (i - 1 >= 0) {
                result += size[group[i - 1][j]]
                set.add(group[i - 1][j])
            }

            if (i + 1 < N && !set.contains(group[i + 1][j])) {
                result += size[group[i + 1][j]]
                set.add(group[i + 1][j])
            }

            if (j - 1 >= 0 && !set.contains(group[i][j - 1])) {
                result += size[group[i][j - 1]]
                set.add(group[i][j - 1])
            }

            if (j + 1 < M && !set.contains(group[i][j + 1])) {
                result += size[group[i][j + 1]]
            }

            answer = max(answer, result)
            set.clear()
        }
    }

    return answer
}

private fun grouping() {
    val visited = Array(N) { BooleanArray(M)}
    var seq = 1


    for (i in 0 until N) {
        for (j in 0 until M) {
            if (board[i][j] == 0 || visited[i][j]) continue
            bfs(i, j, visited, seq++)
        }
    }
}

private fun bfs(row: Int, col: Int, visited: Array<BooleanArray>, seq: Int) {
    val q: Queue<Pair<Int, Int>> = LinkedList()
    visited[row][col] = true
    q.add(Pair(row, col))

    while (q.isNotEmpty()) {
        val (r, c) = q.poll()

        size[seq]++
        group[r][c] = seq

        for (i in 0 until 4) {
            val nr = r + d[i][0]
            val nc = c + d[i][1]

            if (!(nr in 0 until N && nc in 0 until M)) continue
            if (visited[nr][nc] || board[nr][nc] == 0) continue
            visited[nr][nc] = true
            q.add(Pair(nr, nc))
        }
    }
}